### DISCUSSION TITLE
How to properly disconnect from the database

### DISCUSSION THREAD

**Discussion by exscruzme** [GodotNuts/GodotFirebase#336] [Q&A]
Hello! We are using the GodotFirebase plugin in our project to connect with the Firebase Realtime Database. It looks like generally this plugin keeps an open connection to the database and gets pushed updates via the Firebase SSE. 

However, in our case, we do periodically want to disconnect from the database (as we don't need our data to be refreshed often). I was wondering if there is a recommended way to gracefully disconnect / close the Firebase database connection. I had thought that potentially calling `queue_free()` on the database references might release the connection, but we are not sure if that is having the desired effect. 

Any suggestions?

**Comment by WolfgangSenff** [comment]
You _should_ just be able to queue_free the reference. If it's not having the desired effect, what's it doing instead?

**Reply by exscruzme** [reply]
Hi @WolfgangSenff ! sorry for the delayed response. 

It looks like for some reason the connections are staying open. We have a timer on the home screen of our app, and if no new data has been pulled from Firebase, upon expiration of the timer and its signal being fired, we queue free the references to both DB paths that we have. However, when we check the network tab, it looks like they are active connections well past our 30 second timer. 

For context, we are building this app in Godot and currently testing it as a web app (we may later export to iOS and Android, which is why we initially went with Godot). Any thoughts on what's going on here?

Here are some relevant screenshots -- 

the first shows the console log of just a few print statements, the timeout occurring (set to 30 seconds), queue_free of the first db reference, than the second plus the 2 connections shown in the network tab. 

![console and network tab](https://user-images.githubusercontent.com/6991957/224796573-d52b0ffc-0509-4cb0-a5da-41b00ddca138.png)

The second screenshot I took a while later, showing that the references, even though they've been freed in Godot still have open network connections that have been open the entire time the window is open.

![network connection open](https://user-images.githubusercontent.com/6991957/224796797-e9e32c0c-c220-46a0-ade5-3d97aba49229.png)

**Comment by WolfgangSenff** [comment]
Yeah, that will happen - I don't believe the REST API defines the "off" method, which is normally what you'd use to disconnect from those. So we have no way to prevent it from receiving the data, just prevent Godot from interacting with it.

**Reply by exscruzme** [reply]
Makes sense!

**Comment by WolfgangSenff** [comment]
Thinking about this more, there may be a thing we can do to support it better. It might be we need to implement something specifically for this scenario, where it just sends a normal HTTPS request instead of using the server-sent events protocol to try to update it. I think without that protocol in place, it may only send/return the value once. I will have to test this though, and it'll be a while before I can get to it. How urgent is this for you?

**Reply by exscruzme** [reply]
Yeah, that would be a great feature to have! We initially tried to do this just sending REST requests, but also needed the Auth features and liked that your plugin bundles all of the Firebase interactions together, so we wanted to use it. 

It would be awesome to have the one-time push/pull ability with the plugin, and we'd love to use it if you plan on adding the feature! As of right now, this is not a deal breaker for us -- we are launching our app soon, and will be monitoring the open Firebase connections. We would like to stay within the free plan of active connections, since we are just starting out, hence wanting to close the connections when possible. 

Thanks for the quick responses! Looking forward to using this new feature when it's added to the plugin :)

**Comment by shafnaz** [comment]
Is this disconnection issue going to cause problems in lobby game? Suppose a checkers game, where the player connects to room after room, If I `queue_free()` the last `FirebaseDatabaseReference ` before playing the next room, is it going to cause any problems?

**Comment by WolfgangSenff** [comment]
I'm not sure. We do support one-shot gets now though, so you should just use that if you need to disconnect stuff - at least where it makes sense.

**Comment by Firty** [comment]
I have a similar problem. I have a server that is on all day, but every hour I get this error, it does not affect its operation, but it is already bothering. It's as if the connection expires and it tries to reconnect, but the http request is still in use. I believe that right after performing a login, getting what I need, I should terminate the connection. But I didn't quite understand how I can do this...
ERROR: HTTPRequest is processing a request. Wait for completion or cancel it before attempting a new one.
   at: request_raw (scene/main/http_request.cpp:82) - Condition "requesting" is true. Returned: ERR_BUSY

**Reply by WolfgangSenff** [reply]
Do you need it to be connected all day? Or do you just need to periodically get the state of it? I suspect there may be a bug in our flow, since I recently figured out that the check_auth_file function doesn't work the way it should. It might mean you need an await somewhere that isn't being awaited, but since I can't see any of your code, it's hard to say.

**Reply by Firty** [reply]
No, I'm just using FireBase to manage login, download a file from FireStore (my update control) and if there is an update it downloads from Storage. Everything works great, but every hour I get this error. I tried using Auth Logout as it seemed to make sense, but I only get the error that there is no encrypted file. I'm still using version 4.8 for Godot 3.5.3.

**Reply by WolfgangSenff** [reply]
Interesting, okay. Sounds like you may need to yield on something instead. I can't look at the moment, and haven't really looked at 3.x in a while, but hopefully I can help you sometime later today. That is certainly a very annoying thing, sorry about that!

**Reply by Firty** [reply]
I analyzed the auth login flow and found the likely problem. Begin_refresh_countdown() this function has a yield with a timer of 1000 seconds, in my case I am calling login_with_email_and_password twice, the first is in the update system and the second is when the user logs in. Probably the first login is fixing the problem because of the yield. As I don't need to update the token, I will just disable this function and my problem will probably be solved.

**Reply by WolfgangSenff** [reply]
Okay, I was going to suggest you could do that, but wanted to make sure it was okay! Because it will still log out part way through use automatically (on the Firebase side).
